### DISCUSSION TITLE
[consul] types expect generic for functions that don't need it

### DISCUSSION THREAD

**Discussion by OmgImAlexis** [DefinitelyTyped/DefinitelyTyped#61733] [Issues with a @types package]
Types: https://github.com/Floupp/DefinitelyTyped/blob/08a61be703d95f82173de8ce5ae3a660c50e0e84/types/consul/index.d.ts#L367-L371

There seems to be a lot of types written this way, I don't know if I'm misunderstanding how to uses this library with ts or if the types are just written incorrectly.

For example one of the types goes like so.

```ts
services: {
    <TData>(opts: Agent.ServicesOptions, callback: Callback<TData>): void;
    <TData>(callback: Callback<TData>): void;
    <TData>(opts?: Agent.ServicesOptions): Promise<TData>;
};
```

Because of the generic above the following has a return type of `unknown`.

```ts
const services = await consul.agent.services();
```

I get that I could write this using the generic but it feels wrong to have to provide my own interface when we know the return type is always going to match the interface I have.
```ts
interface Service {
    ID: string;
    Service: string;
    Tags: string[];
    Meta: Record<string, unknown>;
    Port: number;
    Address: string;
    SocketPath: string;
    Weights: {
        Passing: number;
        Warning: number;
    };
    EnableTagOverride: boolean;
    Datacenter: string;
}
const services = await consul.agent.services<Service[]>();
```

Would the correct thing be to remove the generics and add this new interface?

[a comment by an automated account is not shown]

**Comment by UNIDY2002** [comment]
You can try out `consul@2.0.0-next.1` which types a great deal of API return values.
